[#297] Add AssertionMatcher and LambdaMatcher

szpak opened the pull request:
The matchers leverages Java 8 to provide inlined ArgumentCaptor with a lambda expression and inlined argument matcher with a predicate lambda expression. More information below and in JavaDoc.


### ~Lambda matcher~

~Allows to define matcher logic within a lambda expression. Useful when working with complex classes pass as an argument.~

Lambda matcher has been removed from the PR as it can be implemented with the enhancements introduced in the master branch in the meanwhile.

### Argument Captor - Java 8 edition

Allows to use ArgumentCaptor in one line:

```
@Test
public void shouldAllowToUseAssertionInLambda() {
    //when
    ts.findNumberOfShipsInRangeByCriteria(searchCriteria);
    //then
    verify(ts).findNumberOfShipsInRangeByCriteria(
        assertArg(sc -> assertThat(sc.getMinimumRange()).isLessThan(2000)));
}
```

In comparison to 3 lines in the classic way:

```
@Test
public void shouldAllowToUseArgumentCaptorInClassicWay() {  //old way
    //when
    ts.findNumberOfShipsInRangeByCriteria(searchCriteria);
    //then
    ArgumentCaptor<ShipSearchCriteria> captor = 
        ArgumentCaptor.forClass(ShipSearchCriteria.class);
    verify(ts).findNumberOfShipsInRangeByCriteria(captor.capture());
    assertThat(captor.getValue().getMinimumRange()).isLessThan(2000);
}
```

Backported from https://github.com/szpak/mockito-java8/
Related to #297.

Btw, I'm not sure if `Primitives` is the best name for a class with `defaultValue()` working also with non primitives classes. I probably liked `HandyReturnValues` more.

szpak commented:
typetools fails with Java 9. Reported an [issue](https://github.com/jhalterman/typetools/issues/34 in ) for that.

ChristianSchwarz commented:
Can you add some sample code to the description? Can you explain the benefit of the new matcher types, why not use a plain ArgumentMatcher?

szpak commented:
There are examples in JavaDoc and on the project webpage: https://github.com/szpak/mockito-java8#provided-add-ons . If you would I can paste them inline.

In short, the syntax is usually much shorter.

TimvdLippe commented on src/main/java/org/mockito/internal/matchers/AssertionMatcher.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public boolean matches(Object argument) {
+        try {
+            assertingLambda.accept((T) argument);
```
Casting here feels wrong. Can we get rid of this?

TimvdLippe commented on build.gradle:
```diff
@@ -75,6 +75,7 @@ dependencies {
 
     provided "junit:junit:4.12", "org.hamcrest:hamcrest-core:1.3"
     compile "org.objenesis:objenesis:2.4"
+    compile 'net.jodah:typetools:0.4.8'
```
Is this library really required, would like to not add a library for 1 method invocation.

szczepiq replied:
Absolutely. No extra libraries please, especially pre 1.0 (no compatibility guarantee).

szpak replied:
Looking at the [code](https://github.com/jhalterman/typetools/blob/master/src/main/java/net/jodah/typetools/TypeResolver.java) there seem to be some corner cases in resolving return types.

In this code it is needed to determine a return type of lambda implementing `Consumer` interface. @raphw Maybe you have some code in ByteBuddy which could be used to do that?

bric3 replied:
Can't we augment our `GenericMetadataSupport` tool ?

TimvdLippe commented on src/main/java/org/mockito/internal/matchers/LambdaMatcher.java:
```diff
@@ -0,0 +1,103 @@
+    @Override
+    public boolean matches(T item) {
+        try {
+            return predicate.test((T)item);
```
Also this cast, this should work just fine right?

szczepiq commented:
Thanks a lot @szpak for pushing this forward!

ArgumentMatcher interface is already a predicate and it can be used as lambda. Why do we need an explicit argLambda method?

Is it possible / good idea to overload argThat() method instead of creating new API methods? (discoverability, clean API).

szpak commented:
I simplified the code and removed `argLambda` and `LambdaMatcher`. I created a new wrapper matcher to be able to provide custom message/description in a case of failure. However, I'm not very satisfied with the name `DescriptableWrapperMatcher`. Do you have any better proposals?

ChristianSchwarz commented on src/main/java/org/mockito/internal/matchers/DescriptableWrapperMatcher.java:
```diff
@@ -0,0 +1,55 @@
+ * @since 3.0.0
+ */
+@Incubating
+public class DescriptableWrapperMatcher<T> implements ArgumentMatcher<T> {
```
@szpak Do we need this class? An other option is to provide an addtional constructor in AssertionMatcher with description parameter.

szpak replied:
I don't see a way to use `AssertionMatcher` here as it accepts a `Cunsumer` while in removed `LambdaMatcher` a `Predicate` argument was required. I simplified it to use just `ArgumentMatcher` interface, however, it doesn't provide a way to store a description (it is just a function argument with one `matches` method (which is equivalent to a `Predicate`).

ChristianSchwarz commented:
@szpak A more general question about this PR: Why should one favor  `assertArg` over an `argThat`. Is the main purpose better readabilty of the error message in case a lamba ArgumentMatcher is used?

```java
//assertArg
verify(ts).findNumberOfShips( assertArg(sc -> assertThat(sc.getMinimumRange()).isLessThan(2000)));

//argThat 
verify(ts).findNumberOfShips( argThat(sc -> sc.getMinimumRange()<2000));
//                                        ^ easier to read condition, cause it is short
```

szpak commented:
@ChristianSchwarz Because of the readability in a case of error (assertion failure). AssertJ generates very meaningful error message (the same as in the ArgumentCaptor used in the old 3-line-long style). With Lambda you only know that "condition was not fulfilled".

TimvdLippe commented:
Instead of relying on run-time type-resolution I would rather solve this problem with a solution that is type-safe. I can probably implement this feature in #742 which should be type-safe?

szpak commented:
@TimvdLippe That's great and it could be another option (or even may deprecate this one in the future), however, it's type safe. It's just a default "empty" value to be returned (to not get runtime error) which is not use externally.

szczepiq commented:
>@ChristianSchwarz Because of the readability in a case of error (assertion failure). AssertJ generates very meaningful error message (the same as in the ArgumentCaptor used in the old 3-line-long style). With Lambda you only know that "condition was not fulfilled".

Good use case. +1

@szpak, why new assertArg() method and not new overload of argThat()?

szpak commented:
@szczepiq As I already replied to Christian:

> ChristianSchwarz Because of the readability in a case of error (assertion failure). AssertJ generates > very meaningful error message (the same as in the ArgumentCaptor used in the old 3-line-long style). > With Lambda you only know that "condition was not fulfilled".

ChristianSchwarz commented:
@szczepiq  
overloading with argThat(Consumer<T>) would be nice, but it doesn't work well with lambdas cause the compiler don't like ambiguity.

szpak commented:
I limited the compatibility of AssertionMatcher to Java 8.

Regarding typetools as a dependency I was considering a repackage of it, however it uses `sun.reflect.ConstantPool` which would lead to compilation errors with Java 9. However, it is just one class with no transitive dependencies. Would it be really be such a problem to use it?

TimvdLippe commented:
I am not in favor of merging this particular implementation of lambda matchers. First of all it requires an extra dependency for this one method, secondly we have no guarantee that it actually works all of the time (no type-safety). I would prefer to create a solution that does not have these two problems, then we can support lambda matchers safely :smile:

szpak commented:
> secondly we have no guarantee that it actually works all of the time (no type-safety)

What do you exactly mean?

TimvdLippe commented:
@szpak [This implementation](https://github.com/mockito/mockito/pull/859/files#diff-ded7a7ed5931f11027ce4f6f43d648f6R80) is not type-safe and relies on run-time type retrieval. As there are bugs with the dependency and therefore type resolution, we have no guarantee we won't run into problems with users.

szpak commented:
I reworked my PR to remove external dependency (one class has been repackaged) and to be Java 9 compatible.

@TimvdLippe The return type of `Consumer<T>` AFAIK cannot be determined in a simpler way. It seems to work (even with Java 9) and even in a case of problems in the future only that one feature would be affected (which can be rewritten on the user side to the classic 3 lines long ArgumentCapture (instead of just one with my solution)). For me it is good to give users a try. This is still an incubating feature.

ChristianSchwarz commented:
@all 
What about renaming `assertArg` to a more generic name?  The interface `Consume<T>` allows to do even more than assertions. 

Some examples:

* Argument manipulation: `when(mock).method(assertArg(queue ->queue.poll ));`
* Argument logging: `verify(mock).method(assertArg(a -> System.out.println(a)));`
* Argument capuring without captor: `verify(mock).method(assertArg(a -> list.add(a)));`

ChristianSchwarz commented on src/test/java/org/mockito/internal/matchers/AssertionMatcherTest.java:
```diff
@@ -0,0 +1,89 @@
+    private ShipSearchCriteria searchCriteria = new ShipSearchCriteria(1000, 4);
+
+    @Test
+    public void shouldAllowToUseArgumentCaptorInClassicWay() {
```
This test can be removed ArgumentCaptors are tested in an other test class.

ChristianSchwarz commented on src/test/java/org/mockito/internal/matchers/AssertionMatcherTest.java:
```diff
@@ -0,0 +1,89 @@
+    @Mock
+    private TacticalStation ts;
+
+    private ShipSearchCriteria searchCriteria = new ShipSearchCriteria(1000, 4);
```
The class `TacticalStation` and  `ShipSearchCriteria` add complexity that isn't needed to test the feature. It should be possible to test assertArg() with a well known interface like `Consumer` or `Function` and a plain `String` or `int` parameter. WDYT?

szpak replied:
In my project I use those two classes also for some other use cases and in fact it can be too complex for just `AssertionMatcherTest`.

ChristianSchwarz commented:
@szpak 

> The return type of Consumer<T> AFAIK cannot be determined in a simpler way.

Maybe this existing solution works:  [GenericTypeExtractor.genericTypeOf(..)](https://github.com/mockito/mockito/blob/release/2.x/src/main/java/org/mockito/internal/util/reflection/GenericTypeExtractor.java)

ChristianSchwarz commented on src/main/java/org/mockito/internal/util/Primitives.java:
```diff
@@ -60,6 +63,17 @@ public static boolean isAssignableFromWrapper(Class<?> valueClass, Class<?> refe
+     * @return the boxed default value for the lambda return type (handling primitive and primitive wrapper types)
+     */
+    public static <T> T defaultValueForConsumerLambda(Consumer<T> consumer) {
+        Class<?>[] typeArgs = TypeResolver.resolveRawArguments(Consumer.class, consumer.getClass());
```
> The return type of Consumer<T> AFAIK cannot be determined in a simpler way.

Maybe the existing [GenericTypeExtractor](https://github.com/mockito/mockito/blob/release/2.x/src/main/java/org/mockito/internal/util/reflection/GenericTypeExtractor.java) works.

```java
Class<?>[] typeArgs =GenericTypeExtractor.genericTypeOf(consumer.getClass(), Object.class, Consumer.class)
```

szpak replied:
Thanks for your suggestion. I checked it and after some modifications (interfaces are considered after classes) I was able to make it work for an anonymous class. Unfortunately it doesn't work with a lambda (Consumer is "an instance of" `ParameterizedType` which is the only supported type).

I've checked also `GenericMetadataSupport` suggested by @bric3, however, it works also only for anonymous classes.

There is probably a possibility to enhance Mockito's internal mechanisms to support both lambdas and Java 9 - see https://github.com/jhalterman/typetools/pull/35/files .

ChristianSchwarz replied:
Maybe @raphw can help here, probably byte-buddy has similar utils.

raphw replied:
Byte Buddy has the `TypeDescription` abstraction which binds type variables down the hierarchy. You can simply navigate down the hierarchy to a given type and then query generic values which are resolved on demand. If this works, I can probably help you to go further from there. Let me know.

ChristianSchwarz replied:
@raphw How to get an `TypeDescription` for a Class-Object? I looked into the byte-buddy code but  don't get it.

raphw replied:
By `TypeDescription.ForLoadedType`.

szczepiq commented:
>@szczepiq overloading with argThat(Consumer<T>) would be nice, but it doesn't work well with lambdas cause the compiler don't like ambiguity.

@ChristianSchwarz, thank you for explaining it. Makes prefect sense now.

>@all
>What about renaming assertArg to a more generic name? The interface Consumer<T> allows to do >even more than assertions.

Absolutely! assertArg is nicely expressive but would look awkward if used not for assertions (and could be even confusing). How about one of the following new methods:

```java
//with
when(mock).method(with(queue ->queue.poll));

//arg
when(mock).method(arg(queue ->queue.poll));

//withArg
when(mock).method(withArg(queue ->queue.poll));
```

ChristianSchwarz commented:
@szczepiq 
> How about one of the following new methods:

- `arg(..)`  feels too close to `argThat(..)` maybe it is confusing for users
- `with(..)` and `withArg(..)` looks both good to me, I may favor the shorter name

TimvdLippe commented:
After reconsideration I am still convinced that the current solution is not good enough. Especially no guaranteed type-safety is a blocker for me. I would therefore like to hold off until we can find a solution that is guaranteed type-safe. I will check if we can have a type-safe solution with #742 where we have the opportunity to redesign these features. WDYT?

szpak commented:
I'm not convinced it is a major issue, but on the other hand I don't want to thrust it forward at all cost. Closing the PR.
